sphinx/util/inspect.py

```diff
@@ -241,14 +239,42 @@ def isclassmethod(
     return False
 
 
+def is_class_method_descriptor(
+    obj: Any, cls: Any = None, name: str | None = None
+) -> TypeIs[types.ClassMethodDescriptorType]:
+    """Check if the object is a class method descriptor type.
+
+    This is typically useful to check if a built-in method is a class method.
+    """
+    if isinstance(obj, types.ClassMethodDescriptorType):
+        return True
+    if cls and name:
+        # trace __mro__ if the method is defined in parent class
+        sentinel = object()
+        for basecls in getmro(cls):
+            meth = basecls.__dict__.get(name, sentinel)
+            if meth is not sentinel:
+                return isinstance(obj, types.ClassMethodDescriptorType)
+    return False
+
+
+def is_class_method_like(
+    obj: Any, cls: Any = None, name: str | None = None
+) -> TypeIs[classmethod | types.ClassMethodDescriptorType]:
```

Comment:
Should we change `isclassmethod` to incorporate `is_class_method_descriptor`? Alternatively, do we need to update the existing call sites of `isclassmethod`?

Reply:
Changing isclassmethod may be annoying for extensions since a classmethod descriptor doesn't have a `__func__` attribute, which makes it incompatible (I tried and our `unwrap_all` failed because of that).

OTOH, we have `ismethoddescriptor` and `isdatadescriptor`, so it's not an issue to also have `isclassmethoddescriptor` (which I would prefer the naming of as it's closer to the existing ones).

Reply:
For the various call sites, I will have a look later (I just woke up!)

Reply:
I added underscores asonceyouconcatenate many words it gets hard to read(!)

Perhaps `is_classmethod_descriptor` (reflecting `classmethod` as one term)?

Reply:
Yes that would be better! IIRC, I think the type is named `classmethod_descriptor` in C as well.